File: April 2025/10.py
# User function Template for python3
import heapq

class Solution:
    def minCost(self, houses):
        def manhattan_distance(p1, p2):
            return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])
        
        n = len(houses)
        if n == 0:
            return 0
        
        min_heap = [(0, 0)]  # (cost, house_index)
        visited = set()
        total_cost = 0
        
        while len(visited) < n:
            cost, house_index = heapq.heappop(min_heap)
            if house_index in visited:
                continue
            visited.add(house_index)
            total_cost += cost
            
            for i in range(n):
                if i not in visited:
                    heapq.heappush(min_heap, (manhattan_distance(houses[house_index], houses[i]), i))
        
        return total_cost

File: April 2025/10.cpp
#include <vector>
#include <queue>
#include <cmath>
#include <unordered_set>

using namespace std;

class Solution {
  public:
    int minCost(vector<vector<int>>& houses) {
        auto manhattanDistance = [](const vector<int>& p1, const vector<int>& p2) {
            return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1]);
        };

        int n = houses.size();
        if (n == 0) return 0;

        // Min-heap: {cost, house_index}
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minHeap;
        vector<bool> visited(n, false);
        int totalCost = 0;

        minHeap.push({0, 0}); // Start with the first house

        while (visited.end() != find(visited.begin(), visited.end(), false)) { // While there are unvisited houses
            auto [cost, houseIndex] = minHeap.top();
            minHeap.pop();

            if (visited[houseIndex]) continue;
            visited[houseIndex] = true;
            totalCost += cost;

            for (int i = 0; i < n; ++i) {
                if (!visited[i]) {
                    int dist = manhattanDistance(houses[houseIndex], houses[i]);
                    minHeap.push({dist, i});
                }
            }
        }

        return totalCost;
    }
};
